### ISSUE TITLE
feature request: idtac_no_newline

### ISSUE THREAD

**Issue by andres-erbsen** [coq/coq#9759]

**Comment by JasonGross** [comment]
This one may be tricky.  AIUI, `idtac` sends back an `infomsg` with no trailing newline (`<infomsg>....</infomsg>`, and the printer takes the infomsg and adds a newline.  You are asking for a version of messages in general that print no trailing newline, and a version of `idtac` which displays such a message?

cc @ejgallego

**Comment by andres-erbsen** [comment]
uh, if it's hard, don't bother, we can keep cutting off the newline after running coq.

**Comment by ejgallego** [comment]
What's the use case?

Indeed what to do with each diagnostic is specific to the backend; for example IDEs will place each located info msg as a list item.

Beware of using the printer as an interface; it is not designed for that, so in the end you will carry a large technical debt; there are better options these days IMHO.

**Comment by andres-erbsen** [comment]
the use case is outputting machinecode from coq. what alternatives would you recommend?

**Comment by ejgallego** [comment]
Wny do you use idtac and not a proper print function `print : foo -> string`?

**Comment by ejgallego** [comment]
We could for example provide an AST reflecting the `Pp.t` type of Coq, so you could properly compose documents.

**Comment by andres-erbsen** [comment]
we have a function like that, the string it outputs is passed into idtac

**Comment by andres-erbsen** [comment]
Pp.t would likely do the job but it sounds like overkill -- I can only foresee using it for concrete strings.

**Comment by ejgallego** [comment]
You can listen to the protocol directly (using sertop) and then you can do what you want with the messages.

> we have a function like that, the string it outputs is passed into idtac

Well, that shouldn't matter too much then, right? In fact you'd like to have a separator, otherwise how do you determine the boundary between two idtacs ? I guess that if you are parsing you don't care so much about the newline, right?

But indeed, that's a bit tricky to control at the tactic level as it is the console printer who adds the newline; it seems reasonable to do so; as for example adding a newline to most messages would for example look bad on CoqIDE [unless we do some post-processing]

**Comment by ejgallego** [comment]
> Pp.t would likely do the job but it sounds like overkill -- I can only foresee using it for concrete strings.

Well `Pp.t` provides you with much more flexibility, is a format akin to HTML so it can achieve much better display.

**Comment by andres-erbsen** [comment]
HTML sounds like the opposite of what we want. I just want to output a bytestring from a .v file dedicated to for that purpose. the output will then be redirected to a file, have chmod+x done to it, and executed.

**Comment by ejgallego** [comment]
So what's your problem then? I am afraid I don't understand it; do you have several calls to idtac? If so, you should use your own composition operator.

**Comment by ejgallego** [comment]
`idtac` is not of type `foo -> string` but of type `foo -> unit`, this is an important difference.

**Comment by JasonGross** [comment]
I believe the problem is that there is, at present, no way to make a `.v` file such that running that `.v` file through `coqc` will print a given string without a trailing newline.

**Comment by andres-erbsen** [comment]
I have one call to idtac. it adds an extra newline at the end of the string passed into it. Currently we cut it out afterwars. I was hoping it would be trivial to have coq not print this newline in the first place. It sounds like it isn't, so feel free to close this issue.

**Comment by ejgallego** [comment]
The thing is that `idtac` doesn't add a newline; the console printer does. That's an important difference. If you listen to the output you will get something such as `(Feedback Info (Pp_string "fooo...."))`

So if you have only call, why is that so important?

If you have several calls then there is a problem in the sense that supporting underlying composition is a tricky business, as in general the printing layer has a hard time guessing. @herbelin proposed to have a flag for this, but I am more of the opinion that users should just program a regular pretty-printer.

As for allowing composition with idtac I guess the easiest part is to use your own composition operator as in
```
foo \
bar
```

**Comment by ejgallego** [comment]
> It sounds like it isn't, so feel free to close this issue

We certainly want to be sure we provide adequate support for pretty printing, however indeed this kind of business tends to be very delicate so all the details should be carefully laid down. In particular the printing sub-system must render well in different output formats [such as tex, html, console, gtk] so spacing assumptions are not so easy to make.

**Comment by andres-erbsen** [comment]
it is "important" because after redirecting stdout I want to end up with a file that does not end in a newline

**Comment by andres-erbsen** [comment]
I think the reason we are not understanding each other here is the you are thinking of outputting a document (which then might be rendered using tex or so on), and I am thinking of outputting *bytes* (just to coqc stdout, without even plausibly supporting rendering of any kind).

**Comment by ejgallego** [comment]
The main problem I have is why are you using `idtac`; that's not meant for printing so it is a horrible hack.

It seems to me that you want `fprintf`, that'd be fine to add.

**Comment by andres-erbsen** [comment]
oh, yeah, my apologies. using idtac is definitely a kludge, it is just currently the closest to what I want (outputting only a newline, not a type annotation (`Check`) or identifier name (`Print`).

**Comment by ejgallego** [comment]
> you are thinking of outputting a document

idtac _outputs_ a document; it is designed that way. It is not a low-level output primitive so indeed you have gonna have all kind of problems using it to print byte-precise stuff.

**Comment by ejgallego** [comment]
I think the best would be to add a `fprintf` primitive indeed, but without more context it is hard to say.

**Comment by andres-erbsen** [comment]
if `fprintf` would need more design, how about something like `puts`? I'm not looking forward to starting a sophisticated design effort just so that I can stop cutting out the last byte in a makefile pipe.

**Comment by SkySkimmer** [comment]
Maybe you should make a plugin, then you can just do whatever makes sense for your use case.

**Comment by ejgallego** [comment]
Better printing facilities do make sense in the core, but there are a bit tricky for a few reasons. For example imagine working on a IDE, and you have some output, the IDE may re-execute them et voilà ! Duplicated stuff.

That requires some thinking, however I do have a proposal à la Jupyter for extraction, IO, etc... that could be OK for most users.

**Comment by ejgallego** [comment]
By the way the plugin is really trivial here indeed, just register to listen to feedback, write as you need.

**Comment by SkySkimmer** [comment]
I was thinking more about something to write a string-typed constr 
directly to a file, without having to extern/mess with notations.

Gaëtan Gilbert

**Comment by gares** [comment]
My impression is that your are using Ltac (and its idtac) to do something it was not designed for. There are no IO primitives, and given that the evaluation order is not very well specified (not to talk about backtraing), I believe IO primitives should not be there.

If you want to print a term in ASCII, I think you should generate an ADT (eg something of type string) and finally extract it to ML where you do have `Printf`.

**Comment by andres-erbsen** [comment]
@gares I call idtac only once, with the entire string in one call. Yes, idtac was not designed for it, but nothing in coq was, and idtac is the closest thing.

**Comment by gares** [comment]
Oh, I thought you wanted to have `idtac "a"; idtac "b" = idtac "ab"`.

> uh, if it's hard, don't bother, we can keep cutting off the newline after running coq.

Then I think this is best.

**Comment by ppedrot** [comment]
Let's close as WONTFIX then.
